processor/src/chiplets/kernel_rom/mod.rs

```diff
@@ -0,0 +1,143 @@
+use super::{BTreeMap, Digest, ExecutionError, Felt, Kernel, TraceFragment, Word, ONE, ZERO};
+use vm_core::chiplets::kernel_rom::TRACE_WIDTH;
+
+#[cfg(test)]
+mod tests;
+
+// KERNEL ROM
+// ================================================================================================
+
+/// Kernel ROM chiplet for the VM.
+///
+/// This component is responsible for validating that kernel calls requested by the executing
+/// program are made against procedures which are contained within the specified kernel. It also
+/// tacks all calls to kernel procedures and this info is used to construct an execution trace of
+/// all kernel accesses.
+///
+/// # Execution trace
+/// The layout of the execution trace of kernel procedure accesses is shown below:
+///
+///   s0   idx   h0   h1   h2   h3
+/// ├────┴─────┴────┴────┴────┴────┤
+///
+/// In the above, the meaning of columns is as follows:
+/// - `s0` is a selector column which indicates whether a procedure in a given row should count
+///   toward kernel access. ONE indicates that a procedure should be counted as a single access,
+///   and ZERO indicates that it shouldn't.
+/// - `idx` is a procedure index in the kernel. Values in this column start at ZERO and are
+///   incremented by ONE for every new procedure. Said another way, if `idx` does not change,
+///   values in `h0` - `h3` must remain the same, but when `idx` is incremented values in `h0` -
+///   `h3` can change.
+/// - `h0` - `h3` columns contain roots of procedures in a given kernel. Together with `idx`
+///   column, these form tuples (index, procedure root) for all procedures in the kernel.
+pub struct KernelRom {
+    access_map: BTreeMap<[u8; 32], ProcAccessInfo>,
+    trace_len: usize,
+}
+
+impl KernelRom {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new [KernelRom] instantiated from the specified Kernel.
+    ///
+    /// The kernel ROM is populated with all procedures from the provided kernel. For each
+    /// procedure access count is set to 0.
+    pub fn new(kernel: &Kernel) -> Self {
+        let mut access_map = BTreeMap::new();
+        for &proc_hash in kernel.proc_hashes() {
+            access_map.insert(proc_hash.into(), ProcAccessInfo::new(proc_hash));
+        }
+
+        Self {
+            access_map,
+            trace_len: kernel.proc_hashes().len(),
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns length of execution trace required to describe kernel ROM.
+    pub fn trace_len(&self) -> usize {
```

Comment:
This could be a `const` fn.

It generally helps to have len functions as `const` because we also can extract these values into the static space, whenever needed

Reply:
Thanks! Fixed!

Reply:
This is a good point, and it would probably be good to go through and update our other len functions to `const` as well. It's pretty trivial, but also maybe better to just do it it while we're thinking about it. Should we make an issue?

Reply:
Yep - let's create an issue.